src/libraries/System.Collections.Immutable/src/System/Collections/Frozen/WrappedDictionaryFrozenDictionary.cs

```diff
@@ -0,0 +1,45 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Runtime.InteropServices;
+using System.Threading;
+
+namespace System.Collections.Frozen
+{
+    internal sealed class WrappedDictionaryFrozenDictionary<TKey, TValue> :
+        FrozenDictionary<TKey, TValue>, IDictionary<TKey, TValue>, IEnumerable<KeyValuePair<TKey, TValue>>
+        where TKey : notnull
+    {
+        private readonly Dictionary<TKey, TValue> _source;
+        private TKey[]? _keys;
+        private TValue[]? _values;
+
+        internal WrappedDictionaryFrozenDictionary(Dictionary<TKey, TValue> source, bool sourceIsCopy) : base(source.Comparer) =>
+            _source = sourceIsCopy ? source : new Dictionary<TKey, TValue>(source);
+
+        /// <inheritdoc />
+        private protected sealed override TKey[] KeysCore =>
+            _keys ??
+            Interlocked.CompareExchange(ref _keys, _source.Keys.ToArray(), null) ??
```

Comment:
Why do we need to ensure the same array is returned here?

Reply:
It's not an immutable surface area, otherwise; multiple calls to the same property could see different objects returned.